shell/platform/linux/fl_basic_message_channel_test.cc

```diff
@@ -0,0 +1,146 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Included first as it collides with the X11 headers
+#include "gtest/gtest.h"
+
+#include "flutter/shell/platform/linux/fl_binary_messenger_private.h"
+#include "flutter/shell/platform/linux/fl_engine_private.h"
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_basic_message_channel.h"
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_standard_message_codec.h"
+#include "flutter/shell/platform/linux/testing/mock_renderer.h"
+
+// Creates a mock engine that responds to platform messages
+static FlEngine* make_mock_engine() {
+  g_autoptr(FlDartProject) project = fl_dart_project_new("data");
+  g_autoptr(FlMockRenderer) renderer = fl_mock_renderer_new();
+  g_autoptr(FlEngine) engine = fl_engine_new(project, FL_RENDERER(renderer));
+  g_autoptr(GError) error = nullptr;
+  EXPECT_TRUE(fl_engine_start(engine, &error));
+  EXPECT_EQ(error, nullptr);
+
+  return static_cast<FlEngine*>(g_object_ref(engine));
+}
+
+static void string_response_cb(GObject* object,
+                               GAsyncResult* result,
+                               gpointer user_data) {
+  g_autoptr(GError) error = nullptr;
+  g_autoptr(FlValue) message = fl_basic_message_channel_send_finish(
+      FL_BASIC_MESSAGE_CHANNEL(object), result, &error);
+  EXPECT_NE(message, nullptr);
+  EXPECT_EQ(error, nullptr);
+
+  EXPECT_EQ(fl_value_get_type(message), FL_VALUE_TYPE_STRING);
+  EXPECT_STREQ(fl_value_get_string(message), "Hello World!");
+
+  g_main_loop_quit(static_cast<GMainLoop*>(user_data));
+}
+
+// Checks can send a message and receive the response
+TEST(FlBasicMessageChannelTest, SendMessage) {
+  g_autoptr(GMainLoop) loop = g_main_loop_new(nullptr, 0);
+
+  g_autoptr(FlEngine) engine = make_mock_engine();
+  FlBinaryMessenger* messenger = fl_binary_messenger_new(engine);
+  g_autoptr(FlStandardMessageCodec) codec = fl_standard_message_codec_new();
+  g_autoptr(FlBasicMessageChannel) channel = fl_basic_message_channel_new(
+      messenger, "test/echo", FL_MESSAGE_CODEC(codec));
+  g_autoptr(FlValue) message = fl_value_new_string("Hello World!");
+  fl_basic_message_channel_send(channel, message, nullptr, string_response_cb,
+                                loop);
+  g_main_loop_run(loop);
```

Comment:
You have expectations in the callbacks, but none in the test methods, which always makes me very nervous; if the channel were to be broken in such a way that it simply didn't call any callbacks, wouldn't this test still pass?

I generally like to add some expectation that explicitly verifies that whatever callbacks I expected to be called actually got called. Normally I do that with lambda capture; here probably the simplest option would be to make user_data a struct that includes a bool that gets flipped.

(This applies to the PR in general, not just this test.)

Reply:
We know the callbacks occurred because they cause the main loop to quit. If they didn't occur the test would never complete. In all the cases I call `g_main_loop_quit()` when the final callback is received.

I was considering if I could/should put in any explicit asserts to confirm the callbacks were made, but it didn't seem necessary and would add a lot more complexity to the test case.

Reply:
Okay, maybe add a comment to that effect in the test cases just before the `g_main_loop_run` line? It was non-obvious to me.